### PR TITLE
ISPN-5988 Fine-/Coarse-grained AtomicMap info is not preserved in cache

### PR DESCRIPTION
**Pull request by pruivo**
https://issues.jboss.org/browse/ISPN-5988

### CODE REVIEW

File: core/src/main/java/org/infinispan/atomic/impl/AtomicHashMap.java [hunk 1]
```diff
@@ -54,35 +53,43 @@
...
+   public static <K, V> AtomicHashMap<K, V> newInstance(Cache<Object, Object> cache, Object cacheKey, ProxyMode proxyMode) {
+      AtomicHashMap<K, V> value = new AtomicHashMap<>(proxyMode);
       Object oldValue = cache.putIfAbsent(cacheKey, value);
       if (oldValue != null) value = (AtomicHashMap<K, V>) oldValue;
       return value;
    }
 
    public AtomicHashMap() {
-      this.delegate = new FastCopyHashMap<K, V>();
+      this(new FastCopyHashMap<>(), ProxyMode.NONE);
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/atomic/impl/AtomicHashMap.java]: I'm not sure the `NONE` option is necessary: I removed the enum value and replaced it here with `COARSE`, and all the tests passed.

File: core/src/main/java/org/infinispan/atomic/impl/AtomicHashMap.java [hunk 2]
```diff
@@ -254,4 +277,9 @@ public Integer getId() {
          return Util.<Class<? extends AtomicHashMap>>asSet(AtomicHashMap.class);
       }
    }
+
+   public enum ProxyMode {
+      NONE, FINE, COARSE;
+      public static final ProxyMode[] CACHED_VALUES = values();
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/atomic/impl/AtomicHashMap.java]: The same comment as for https://github.com/infinispan/infinispan/pull/3930, having this field public is confusing.

File: core/src/main/java/org/infinispan/atomic/impl/AtomicHashMap.java (hunk 1 above)
Comment (pruivo) [inline core/src/main/java/org/infinispan/atomic/impl/AtomicHashMap.java]: because that constructor is no longer used. I'll remove it.

Comment (pruivo) [inline core/src/main/java/org/infinispan/atomic/impl/AtomicHashMap.java]: edit: actually, it is used in tests :)

**Comment by danberindei** [comment]
Integrated, thanks Pedro!

**Pull request closed without merging**